File: sdk/lua/sdk/AbstractDumper.lua
local IDumper = {}
IDumper.__index = IDumper

function IDumper:getRoot() 
    -- :rettype: support.poco.sdk.AbstractNode
end

function IDumper:dumpHierarchy() 
    -- :rettype: dict or NoneType
end

local AbstractDumper = {}
AbstractDumper.__index = AbstractDumper
setmetatable(AbstractDumper, IDumper)


function AbstractDumper:dumpHierarchy() 
    return self:dumpHierarchyImpl(self:getRoot())
end


function AbstractDumper:dumpHierarchyImpl(node, onlyVisibleNode)
    if node == nil then
        return nil
    end
    if onlyVisibleNode == nil then
        onlyVisibleNode = true
    end

    local payload = node:enumerateAttrs()
    local result = {}
    local children = {}
    for _, child in ipairs(node:getChildren()) do
        if not onlyVisibleNode or (payload['visible'] or child:getAttr('visible')) then
            table.insert(children, self:dumpHierarchyImpl(child, onlyVisibleNode))
        end
    end
    if #children > 0 then
        result['children'] = children
    end
    
    result['name'] = payload['name'] or node:getAttr('name')
    result['payload'] = payload
    
    return result
end

return AbstractDumper

File: sdk/js/sdk/AbstractDumper.js
var IDumper = function () {

}

IDumper.prototype.getRoot = function () {
    // :rettype: support.poco.sdk.AbstractNode
}

IDumper.prototype.dumpHierarchy = function () {
    // :rettype: dict or NoneType
}

var AbstractDumper = function () {
    IDumper.call(this)
}
AbstractDumper.prototype = Object.create(IDumper.prototype)

AbstractDumper.prototype.dumpHierarchy = function () {
    return this.dumpHierarchyImpl(this.getRoot())
}

AbstractDumper.prototype.dumpHierarchyImpl = function (node, onlyVisibleNode) {
    if (!node) {
        return null
    }
    if (onlyVisibleNode === undefined) {
        onlyVisibleNode = true
    }

    var payload = node.enumerateAttrs()
    var result = {}
    var children = []
    var nodeChildren = node.getChildren()
    for (var i in nodeChildren) {
        var child = nodeChildren[i]
        if (!onlyVisibleNode || (payload['visible'] || child.getAttr('visible'))) {
            children.push(this.dumpHierarchyImpl(child, onlyVisibleNode))
        }
    }
    if (children.length > 0) {
        result['children'] = children
    }

    result['name'] = payload['name'] || node.getAttr('name')
    result['payload'] = payload

    return result
}

try {
    module.exports = AbstractDumper;
} catch (e) {
    if (window.module && window.module.exports) {
        window.module.exports = AbstractDumper;
    }
}